conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
+                            assert n_sparse_dims <= len(factors)
+                            flat_indices0 = torch.zeros(
+                                n_values, dtype=torch.int64, device=d_p.device)
+                            for i_sparse_dim in range(n_sparse_dims):
+                                flat_indices0.add_(
+                                    factors[i_sparse_dim],
+                                    d_p._indices()[i_sparse_dim])
+                            flat_indices0.unsqueeze_(1)
+                            offsets = torch.arange(value_size, 
+                                dtype=torch.int64, device=d_p.device).\
+                                unsqueeze(0)
+                            flat_indices = \
+                                torch.index_select(
+                                    flat_indices0, 1,
+                                    torch.zeros(
+                                        value_size, 
+                                        dtype=torch.int64,
+                                        device=d_p.device)) \
+                                + \
+                                torch.index_select(
+                                    offsets, 0,
+                                    torch.zeros(
+                                        n_values,
+                                        dtype=torch.int64,
+                                        device=d_p.device))
+                            flat_indices.resize_(values_shape)
+                            del flat_indices0, offsets
+                            ctor = eval(d_p.type())
+                            vals = torch.take(p.data, flat_indices)
+                            try: 
+                                sparse_data = ctor(
+                                    d_p._indices(), vals, d_p.size())
+                            except RuntimeError:
+                                # This looks like a bug in pytorch; sparse
+                                # tensors evidently cannot be created on
+                                # every gpu, even if all the data is on the
+                                # same gpu.  The workaround:  construct the 
+                                # sparse tensor on the cpu then move it.
+                                ctor_cpu = eval(
+                                    re.sub('\.cuda', '', d_p.type()))
+                                sparse_data = ctor_cpu(
+                                    d_p._indices().cpu(), vals.cpu(),
+                                    d_p.size()).to(d_p.device)
+                            d_p.add_(weight_decay, sparse_data)
+                            del flat_indices, sparse_data, vals
+                if momentum != 0:
+                    param_state = self.state[p]
+                    if 'momentum_buffer' not in param_state:
+                        buf = param_state['momentum_buffer'] = torch.zeros_like(d_p)
+                        buf.mul_(momentum).add_(d_p)
```

Comment:
It would seem that something interesting is going on with sparse matrix internals that's not evident from the code. What are these `mul_` and `add_` methods? I can't even Google them to increase my understanding.

Reply:
Sparse tensors inherit these methods from torch.Tensor (documented on https://pytorch.org/docs/stable/sparse.html and  https://pytorch.org/docs/stable/tensors.html).  As usual in PyTorch, the trailing underscore indicates a method that modifies its object in place, and mul and add refer to scalar multiplication and tensor addition.